### ISSUE TITLE
cannot import name 'Adapter' from 'pylogix' 

### ISSUE THREAD

**Issue by meenavinod** [dmroeder/pylogix#207]
### Preflight checks

Before you post an issue, ensure you have tried the minimal examples within the repo, or tried the pylogix-tester.

- [ ] Have you tried the [examples](https://github.com/dmroeder/pylogix/tree/master/examples)?
- [ ] Have you tried [pylogix-tester](https://github.com/TheFern2/pylogix-tester)?
- [ ] Have you read previous issues?

## Type of issue
Import Error 
- [ ] Bug
- [ ] Feature Request
- [ ] Question
- [ ] Other

> Delete items that do not apply below.

## Description of issue
Unable to Start Pylogix Adaptor 
## Code

Please provide a minimal, reproducible example

## Versions

Include versions to

- pylogix:
- python:
- OS:

**Comment by TheFern2** [comment]
Pylogix does not have an Adapter class to my knowledge, on which example did you see Adapter being imported?

- What are you trying to do?

**Comment by dmroeder** [comment]
Pylogix does have an Adapter class, but only in the experimental branch.  Since this is not in the main branch, you must download the project from here on GH, then switch to the experimental branch.

**Comment by TheFern2** [comment]
Ah ooppss, sorry for the wrong information, wasn't aware of it until now.

**Comment by TheFern2** [comment]
Did you posted a comment and accidentally deleted? I see it on the notifications but is not here.

**Comment by meenavinod** [comment]
Hi,
Thanks for very much for quick responses. I tried the above suggestions and finally this issue looks fine but I am still unable to communicate with  PLC. I am still facing below issues - 
1. Unable to restore L5X file to ACD.
![image](https://user-images.githubusercontent.com/116361822/199458675-e85e56e5-65a1-4b51-befe-720a0cd8e858.png)

2. I tried to incorporate L5X changes to my test project but still no connectivity with adapter
![image](https://user-images.githubusercontent.com/116361822/199459267-14612724-7fba-4a9b-9c5f-5c505fa03a0a.png)
Can you please provide the ACD file with example..?

**Comment by meenavinod** [comment]
I tried to import L5X IO adapter as ENBT child module. I  made some correction to add as a remote child module of ENBT. My project IO tree have adaptor module as above. 
Please let me know if I did anything wrong..?

**Comment by dmroeder** [comment]
Logix is probably being picky because I exported the module from a CompactLogix.  I'm guessing that it doesn't like importing to the ENBT.  Your image above doesn't tell me how you configured the module.  The example expects that you configured it like the image:
https://github.com/dmroeder/pylogix/blob/experimental/examples/100_simple_adapter.png

The adapter in the I/O tree needs the IP address of the machine that will be running the pylogix adapter.  And your adapter script needs to be configured for your ENBT's IP address.  If that is not working, post a screenshot of your I/O adapter configuration, you exact python script and your IP adddresses.

**Comment by meenavinod** [comment]
Hi, When i tried to import L5X to my project file, its giving error of parent module 
![image](https://user-images.githubusercontent.com/116361822/199640448-d26ec865-7145-4f5e-b270-44a8ca57f350.png)

I believe adapter cant be added as Local, so I changed the parent module properties in L5X file and import again as child of ENBT 
![image](https://user-images.githubusercontent.com/116361822/199640834-231b557c-9e48-476c-bb6b-7fc990cb9f12.png)

I used same IP address as per example and used same script to run the adapter 
![image](https://user-images.githubusercontent.com/116361822/199640982-dd7a3456-beca-4be4-8096-ec0dcb6ac244.png)
![image](https://user-images.githubusercontent.com/116361822/199641091-cf61e8db-55d4-47b6-95b7-3f2e824e1e14.png)

Still PLC and Adapter unable to connect...

**Comment by dmroeder** [comment]
As I mentioned, Logix is not letting you import the module because it is terrible.  I had the module in a CompactLogix, you have a ControlLogix, it is not smart enough to disregard that difference, unfortunately :)

Is your computer's network adapter configured for the IP Address 192.168.1.23?  Is your ENBT's IP Address 192.168.1.10?

**Comment by meenavinod** [comment]
Yes, IP is same as above. Its able to detect hostname correctly ...
![image](https://user-images.githubusercontent.com/116361822/199642748-cdf82aa0-d25e-4a09-9bb1-e96427b66fde.png)

**Comment by dmroeder** [comment]
Can you capture it via wireshark and email the result to me?

I do have a ControlLogix and ENBT, though I won't be able to test until Friday.

**Comment by meenavinod** [comment]
![image](https://user-images.githubusercontent.com/116361822/199644736-0dfd325b-d02e-4daa-b39b-2bf928bc3252.png)

**Comment by dmroeder** [comment]
Is your script running in that instance?   All I see is list identity, which is not pylogix, it is probably RSLinx.   To minimize wireshark traffic, shut down RSLinx and LogixDesigner.  Run Wireshark, run your script.

**Comment by meenavinod** [comment]
![image](https://user-images.githubusercontent.com/116361822/199647910-210f3c3f-1e32-442b-b36c-309dfea4c52c.png)

Without RS Linx and Adapter Script running

**Comment by dmroeder** [comment]
Here is the problem I'm having.  You show the screenshot of your generic module's Module Info tab, it is reporting a Rockwell Software Communication Adapter.  You are telling me that your computers IP address is 192.168.1.23 and that generic module is configured for 192.168.1.23, which looks to be the case and that would be correct. 

There's something up with your screenshot though.  When you open the Module Info of your generic module, it is sending out a List Identity request to the address of 192.168.1.23.  Some Rockwell device is responding.  If it is your computer that is configured for 192.168.1.23, and it should be based on your script, nothing should respond.  The only thing that should be at .23 is your computer, which would not respond to a List Identity request.

**Comment by dmroeder** [comment]
Wait, can you go to the Network and Sharing Center and check if your adapter is set to "public network"?

I had trouble with this, the firewall was blocking the connect request from the PLC.

**Comment by meenavinod** [comment]
I believe its reporting a Rockwell Software Communication Adapter because of RS Linx. If I change the IP address of my computer its unable to find module properties -
![image](https://user-images.githubusercontent.com/116361822/200099527-a3a80d2d-52c1-4b88-9041-7dc14e507f43.png)

My network and sharing center already was in Public network but Firewall Defender was on. After disabling it After disabling the Firewall defender for Public network -
![image](https://user-images.githubusercontent.com/116361822/200099214-f4ff6c10-5379-437d-93bd-03e5d4b4f297.png)

**Comment by dmroeder** [comment]
Okay, you are getting close.  I don't think RSLinx would cause what you were seeing but maybe RSlinx Enterprise, I can't say for sure.  The timeout is what I would expect because the pylogix adapter hasn't implemented the list identity response.

You may have some more firewall work to do.  I think there should be a register session, which I don't see.  After both the register session and forward open, the connection switches to UDP, port 2222.

**Comment by meenavinod** [comment]
I Think its registering the session but Forward Open connection failed...

I have disabled firewall completely .

![image](https://user-images.githubusercontent.com/116361822/200104790-1c2b8825-cf60-4bad-9764-8c0a32c0fc42.png)

**Comment by dmroeder** [comment]
Can you email me that wireshark packet capture?

**Comment by meenavinod** [comment]
Hi, You are right, I didnt noticed any change with Adapter Script running or not in Wireshark. Can you please share your mail address. my email is meenavinod017@yahoo.com.
![image](https://user-images.githubusercontent.com/116361822/200219317-c29ab2d6-64d9-420f-a9c4-f465021f5d4e.png)
 
![image](https://user-images.githubusercontent.com/116361822/200219367-056060bd-5c0f-44cc-9489-a487166174f7.png)

![image](https://user-images.githubusercontent.com/116361822/200219431-342d7ae2-6270-4bdd-b95d-32645f42e572.png)

But the Pylogix eip working as expected with same setup. 
![image](https://user-images.githubusercontent.com/116361822/200219757-2c977632-a28d-4b0b-989d-c9c35425425e.png)

**Comment by lbhopper** [comment]
have you tried disabling keys in your Adapter Module? ControlLogix modules can return with errors if keying is set to exact or compatible:

https://literature.rockwellautomation.com/idc/groups/literature/documents/at/logix-at001_-en-p.pdf

**Comment by nucleonautomation** [comment]
I built something for Ethernet Ip adapter, the adapter is in form of binaries and can be used for R&D and lab projects, or for fun 🤩 

Github: https://github.com/nucleonautomation/EnIP-Adapter

Webpage: https://nucleonautomation.com/service/enip-adapter

It has all services and classes inbuilt and a very minimal API for python.
